### PR TITLE
Call old assumptions on Expr instead of Symbol

### PR DESCRIPTION
**Pull request by debugger22**

### CODE REVIEW

**Comment by debugger22** [comment]
@asmeurer It's working now!

File: sympy/assumptions/assume.py [hunk 1]
```diff
@@ -176,6 +176,10 @@ def eval(self, expr, assumptions=True):
                 except AttributeError:
                     continue
                 res = eval(expr, assumptions)
+                # Do not stop if Expr method returned None
+                # Try to check for higher classes
```
Comment (asmeurer) [inline sympy/assumptions/assume.py]: Shouldn't this logic be applied to all classe?

Comment (debugger22) [inline sympy/assumptions/assume.py]: The `Expr` method prevented the mro from evaluating `Basic`. I don't understand why it should be applied to all.

Comment (asmeurer) [inline sympy/assumptions/assume.py]: It should always walk up the class hierarchy until it gets a non-None answer.

**Comment by debugger22** [comment]
Done. But that made it a little slower.

master
```
________________________________ slowest tests ________________________________
test_zero - Took 10.072 seconds
test_zero_positive - Took 11.136 seconds
test_oddness_in_ternary_integer_product_with_even - Took 12.012 seconds
test_issue_7246 - Took 13.275 seconds
test_rational_irrational - Took 14.426 seconds
test_algebraic - Took 15.913 seconds
test_integer - Took 15.976 seconds
test_evenness_in_ternary_integer_product_with_odd - Took 19.305 seconds
test_oddness_in_ternary_integer_product_with_odd - Took 21.248 seconds
test_positive - Took 24.858 seconds
test_negative - Took 26.377 seconds
test_hermitian - Took 27.662 seconds
test_even - Took 34.158 seconds
test_odd - Took 59.144 seconds
test_imaginary - Took 93.850 seconds
test_real - Took 119.999 seconds
test_pow - Took 140.917 seconds
```

this branch
```
test_evenness_in_ternary_integer_product_with_even - Took 10.214 seconds
test_rational - Took 11.008 seconds
test_zero_positive - Took 11.301 seconds
test_nonzero - Took 11.971 seconds
test_oddness_in_ternary_integer_product_with_even - Took 12.297 seconds
test_issue_7246 - Took 13.310 seconds
test_rational_irrational - Took 14.461 seconds
test_integer - Took 15.959 seconds
test_algebraic - Took 19.285 seconds
test_evenness_in_ternary_integer_product_with_odd - Took 20.255 seconds
test_zero - Took 21.133 seconds
test_oddness_in_ternary_integer_product_with_odd - Took 23.531 seconds
test_positive - Took 25.237 seconds
test_negative - Took 26.238 seconds
test_hermitian - Took 27.432 seconds
test_even - Took 35.735 seconds
test_odd - Took 60.629 seconds
test_imaginary - Took 94.146 seconds
test_real - Took 134.222 seconds
test_pow - Took 145.633 seconds
```

**Comment by asmeurer** [comment]
Do you think it's a problem?

**Comment by debugger22** [comment]
No, it just increases few more calls.

**Pull request merged**